3016 integrate paypal api workflow

mattwr18 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
this works for signing up Premium members through the PayPal rest SDK
needs to be set up to dynamically create plans based on premium, premiummob, premiumf2f
also, needs to udpate the backend, handle sponsorships, etc.
Also, it is really slow, so need to find a way to optimize it... it would be nice to encorporate JavaScript as well.
#### Issue addressed
<!-- include `fixes #<issue-number>` -->
<!-- the relevant issue in https://github.com/AgileVentures/WebsiteOne/issues  -->

fixes #3016 

#### Screenshots (if appropriate):
<!-- please include screenshots of any changes to the UI for quick review  -->
<!-- please show how things look on both desktop and mobile  -->
![screenshot at 2019-01-24 00 10 31](https://user-images.githubusercontent.com/26943915/51649653-7ddbda00-1f6c-11e9-8530-5d3c63dfa0e8.png)

tansaku commented on app/controllers/subscriptions_controller.rb:
```diff
@@ -96,11 +96,7 @@ def detect_plan_after_payment
-           else
-             current_user
-           end
+    user = params[:user] ? User.find_by(slug: params[:user]) : current_user
```
nice

tansaku commented on app/controllers/subscriptions_controller.rb:
```diff
@@ -96,11 +96,7 @@ def detect_plan_after_payment
-             current_user
-           end
+    user = params[:user] ? User.find_by(slug: params[:user]) : current_user
     User.find(user.id)
```
is this redundant now? could we even change to:

```rb
return User.find_by slug: params[:user] if params[:user]
current_user
```

or even

```rb
User.find_by(slug: params[:user] || current_user.slug)
```

sorry, I've been gripped by refactoring madness!  that last one is silly - the two liner above probably best ...

tansaku commented on app/services/paypal_service.rb:
```diff
@@ -1,7 +1,7 @@
 
 PayPal::SDK::REST.set_config(
-  mode: 'sandbox', # "sandbox" or "live"
+  mode: Rails.env.production? ? 'live' : 'sandbox',
```
one thing to note here is that I don't think we should switch to live based on the production mode of the rails app - sometimes we might run production locally, or we might be on staging or develop servers (which run in production mode) and we still want to use the sandbox.

maybe better to have an ENV var that can be set explicitly so that the default in any mode is sandbox, but so we can set PAYPAL_SDK_MODE to 'live' on the production server and only there

tansaku commented:
really great work here @mattwr18 !

tansaku commented:
thanks @mattwr18 - looking good

wonder what's up with the code coverage?  wonder if the cucumber is being properly included there ...

tansaku commented:
presumably they are both exercised by the cucumber tests? maybe it is expecting rspec coverage ...

tansaku commented:
@mattwr18 at the moment codeclimate seems happy, but the build is failing ...

mattwr18 commented:
yes, hence why I asked should I be going back two commits when the build was passing, but code climate wasn't

the last time we talked about this you said that maybe I misunderstood what you were getting at and started writing rspec tests

tansaku commented:
@mattwr18 apologies - I hadn't understood that the codeclimate was failing on a previous build - I had lost the context when I was reviewing.

I guess it's either rspec tests OR ensure that the cucumber tests are properly contributing to the coverage, OR just ignoring the coverage - judgement call.  Do you have an angle on how to investigate whether or not the cucumber tests are contributing to the coverage?

mattwr18 commented:
It looks like we are using the `features` directory in the `.simplecov` @tansaku 
https://github.com/AgileVentures/WebsiteOne/blob/3016-integrate-paypal-api-workflow/.simplecov#L2

tansaku commented:
@mattwr18 looks like the coverage is good now ... but the build is failing:

```
Failures:

  1) PaypalAgreementController POST /paypal/new posts to the api without crashing
     Failure/Error: expect(response).to redirect_to(redirect_url)
       Expected response to be a <3XX: redirect>, but was a <500: Internal Server Error>
     # ./spec/requests/paypal_agreement_spec.rb:23:in `block (3 levels) in <main>'

Finished in 43.39 seconds (files took 2.71 seconds to load)
665 examples, 1 failure

Failed examples:

rspec ./spec/requests/paypal_agreement_spec.rb:18 # PaypalAgreementController POST /paypal/new posts to the api without crashing
```

mattwr18 commented:
yes @tansaku we want to finish this off during the `Rails 5 TDD Mob`, we are having difficulty at the moment with stubbing out the PayPal API call :cry:  

we had a look last week, but it's still making a API call when we call `post `/payal/new`

mattwr18 commented:
we had another look today in the `Rails 5 TDD Mob` and we are still blocked the proper stubbing of the external `API` call. 

maybe we need to be using vcr @tansaku? or we need to have a better look in the meeting tomorrow, cause this `PR` is getting old

mattwr18 commented:
thanks to @sherspock, we got passed the calling the api, and now we are stuck at 
NoMethodError for `error`, and `links`

tansaku commented on app/controllers/paypal_agreement_controller.rb:
```diff
@@ -0,0 +1,32 @@
+class PaypalAgreementController < ApplicationController
+  def new
+    @plan = Plan.find(params[:plan])
+    binding.pry
```
@mattwr18 was just trying to check travis for the latest error, but saw you have  a binding.pry in here - part of the latest round of debugging.  I'll remove locally and check, but I was just wondering if the latest code you and Ryder were looking at is in the branch, or do you have some additional change locally that I should have before I start my review?

mattwr18 replied:
that's the latest code @tansaku, I thought I had removed all the `binding.pry`s, but they are sneaky sometimes.

there are some things that I know need to be changed, like the `redirect_url` from here https://github.com/AgileVentures/WebsiteOne/pull/3018/files#diff-8ba20331054f486b79d6002c4ebd45a1R6 does not match the one from the fixtures here https://github.com/AgileVentures/WebsiteOne/pull/3018/files#diff-bb17bf70169813026eb3e62c87ef480dR40, but that is not what I am blocked by

tansaku commented:
Running locally I get this error in the specs:

```
Failures:

  1) PaypalAgreementController POST /paypal/new redirects to a redirect url on successful agreement creation
     Failure/Error: let(:premium_plan) { FactoryBot.create(:plan, name:'premium', paypal_id: 'P-6A8554404E294635VUDTNLWQ') }
     
     NoMethodError:
       undefined method `paypal_id=' for #<Plan:0x00007fa540b8e808>
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/activemodel-5.2.2/lib/active_model/attribute_methods.rb:430:in `method_missing'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/factory_bot-4.11.1/lib/factory_bot/attribute_assigner.rb:16:in `public_send'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/factory_bot-4.11.1/lib/factory_bot/attribute_assigner.rb:16:in `block (2 levels) in object'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/factory_bot-4.11.1/lib/factory_bot/attribute_assigner.rb:15:in `each'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/factory_bot-4.11.1/lib/factory_bot/attribute_assigner.rb:15:in `block in object'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/factory_bot-4.11.1/lib/factory_bot/attribute_assigner.rb:14:in `tap'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/factory_bot-4.11.1/lib/factory_bot/attribute_assigner.rb:14:in `object'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/factory_bot-4.11.1/lib/factory_bot/evaluation.rb:13:in `object'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/factory_bot-4.11.1/lib/factory_bot/strategy/create.rb:9:in `result'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/factory_bot-4.11.1/lib/factory_bot/factory.rb:43:in `run'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/factory_bot-4.11.1/lib/factory_bot/factory_runner.rb:29:in `block in run'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/activesupport-5.2.2/lib/active_support/notifications.rb:170:in `instrument'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/factory_bot-4.11.1/lib/factory_bot/factory_runner.rb:28:in `run'
     # /Users/tansaku/.rvm/gems/ruby-2.5.1/gems/factory_bot-4.11.1/lib/factory_bot/strategy_syntax_method_registrar.rb:20:in `block in define_singular_strategy_method'
     # ./spec/requests/paypal_agreement_spec.rb:5:in `block (3 levels) in <main>'
     # ./spec/requests/paypal_agreement_spec.rb:10:in `block (3 levels) in <main>'

Finished in 1 minute 10.2 seconds (files took 21.57 seconds to load)
665 examples, 1 failure

Failed examples:

rspec ./spec/requests/paypal_agreement_spec.rb:9 # PaypalAgreementController POST /paypal/new redirects to a redirect url on successful agreement creation

Randomized with seed 31659
```

I'm running cukes as well to check - but is this the error you were working on @mattwr18 ?

mattwr18 commented:
this is the test, but not the error message I was getting, have you run the migrations?
`db/migrate/20190129191454_add_paypal_id_to_plans.rb`

tansaku commented:
migrations are run:

```
   up     20181220160421  Create join table project slack channel
   up     20190129191454  Add paypal id to plans
   up     20190312165636  ********** NO FILE **********
```

I had the branch locally from before.  premium cukes are all green ...

mattwr18 commented:
I'm not sure then @tansaku... I have the ability to set the `Plan.first.paypal_id` locally

This `rails db:migrate:status` is for `RAILS_ENV=test`? sorry, don't mean to ask you these questions, I'm sure you have thought of that, but I am not sure why it would behave differently on you machine, and maybe you just forgot?

tansaku commented:
@mattwr18 - yes it's for RAILS_ENV=test - sorry I should have included the full command to make it clear that I was indeed looking at the test database:

```
$ rake db:migrate:status RAILS_ENV=test

database: websiteone_test

 Status   Migration ID    Migration Name
--------------------------------------------------
...
   up     20181220160421  Create join table project slack channel
   up     20190129191454  Add paypal id to plans
   up     20190312165636  ********** NO FILE **********
```
it's entirely possible there's all sorts of other junk on my machine that's causing this.  I'll likely drop the database (and do other clear-up) as a next step when I get a chance to explore further.

What I was hoping you might do in the meantime is push up code with the `binding.pry` removed so that we can see how it performs in the CI.  Of course I could remove the `binding.pry` myself, but I was concerned that there might be other changes you had locally that might not yet be in the branch on github.  If you did already check that and there's nothing missing, then of course I can remove the `binding.pry` when I next push something up and I'll hopefully be able to get to the details of

> NoMethodError for error, and links

Or I may get to them with some more cleaning up on my side, so no rush, but ultimately I'm hoping to see details of the error you're seeing locally both in CI and my own machine.  I'll try to make more time to look at this later today.

mattwr18 commented:
hey @tansaku, I've removed the `binding.pry` and pushed it up just to make sure you have the latest.
here are some screenshots of the rspec failure message, and where the error is occurring
![Screenshot at 2019-05-09 09:32:01](https://user-images.githubusercontent.com/26943915/57453791-acd93b80-723d-11e9-821c-1b3f304918dd.png)
![Screenshot at 2019-05-09 09:31:49](https://user-images.githubusercontent.com/26943915/57453797-b06cc280-723d-11e9-87f1-914f41f06d17.png)

mattwr18 commented:
it is trying to call the method `error` on what I have returned from the `create_agreement` method, which is a `String`. 
Is there a way we can stub it so that it returns a valid `PayPal::SDK::REST::DataTypes::Agreement` object @tansaku? 
I was looking at the `vcr` using in the rspec tests, but unless I am missing something these wouldn't respond to `@agreement.error`/`@agreement.links.find` either

tansaku commented:
thanks for removing the binding.pry.  Getting the expected error in the build:

```

Failures:

  1) PaypalAgreementController POST /paypal/new redirects to a redirect url on successful agreement creation
     Failure/Error: expect(response).to redirect_to(redirect_url)
       Expected response to be a <3XX: redirect>, but was a <500: Internal Server Error>
     # ./spec/requests/paypal_agreement_spec.rb:12:in `block (3 levels) in <main>'

Finished in 48.44 seconds (files took 3.37 seconds to load)
665 examples, 1 failure

Failed examples:

rspec ./spec/requests/paypal_agreement_spec.rb:9 # PaypalAgreementController POST /paypal/new redirects to a redirect url on successful agreement creation
```

will try to replicate locally ...

tansaku commented:
@mattwr18 so I fixed it locally.  It's actually a tricky one.  The usual methods of turning a hash into an object using OpenStruct or https://github.com/intridea/hashie can't be used by themselves since `method` is a ruby keyword, so I ended up creating a custom object; we'll see if it passes on CI.

I'm interested to know your motivation for creating a spec on this particular controller.

mattwr18 commented:
hey, it passed the build!!
the reason I added these was because the code coverage was failing saying we weren't testing the `PaypalAgreementController` or `PaypalService`
I think the cucumber tests are testing just the `subsriptions_controller`, after the successful agreeement is created.

mattwr18 commented:
@tansaku, not sure if we want to do this or not, but I attempted to add more of a round trip by simulating a redirect from paypal to the `PayPalAgreementController#create` and expecting that to create the agreement, then redirect to `Subscription#create`.

I added a `VCR`, filtered the sensitive data, and was able to successfully create the agreement, but for some reason it is not redirecting to `Subsription#create`, so an email is not being sent out. I added a `binding.pry` and it is not reaching that part of the code. Maybe something jumps out at you?

I can also revert the code, if you think it's not worth it...

mattwr18 commented:
hey @tansaku, got the build passing by following your advice to use `visit` instead of `get` and tweaking some things... I didn't modify the `sponsor_other_user` to pass through the `paypal_agreement_controller` since the implementation relies on setting the session to the sponsored user and passing it as params to the `subscriptions_controller` and this is non-trivial in capybara.

I did find this https://stackoverflow.com/questions/21491530/cucumber-and-capybara-scenario-session, which talks about how it might be done by adding a gem `rack_session_access`.

The codeclimate is still failing, so not really sure what to do about that...

tansaku commented:
@mattwr18 great work here - I've fixed the code climate code issues - and the only things remaining are minor code coverage changes.

Code climate is not giving us the details stats to fix this.  I think we need a ticket on getting some good dev machine code coverage during development, so I'll merge this now.  Great work here! :-)

tansaku commented:
I just created #3214

mattwr18 commented:
this is great @tansaku! thanks for merging it, sorry it took so long to get it merge-able...
this does require some tasks before it will work in production that will require our coordination, I believe
we need to create the plans on AgileVentures' PayPal since I've been using it on my own.
we also need to add these plan id's to the `Plan` as `paypal_id` in our production database, which I can do on deploy day...
